spectral_cube/spectral_cube.py

```diff
@@ -1912,6 +1912,68 @@ def subcube_from_ds9region(self, ds9region, allow_empty=False):
         # that we can now crop out
         return masked_subcube.minimal_subcube(spatial_only=True)
 
+    def subcube_from_ds9region_new(self, ds9region, allow_empty=False):
+        """
+        Extract a masked subcube from a ds9 region or a Region list object
+        (only functions on celestial dimensions)
+
+        Parameters
+        ----------
+        ds9region: str or `regions.Region` list
+            The region to extract
+        allow_empty: bool
+            If this is False, an exception will be raised if the region
+            contains no overlap with the cube
+        """
+        import regions
+
+        if isinstance(ds9region, six.string_types):
+            shapelist = regions.DS9Parser(ds9region).shapes.to_regions()
+        else:
+            shapelist = ds9region
+
+        regionlist = []
+        for x in shapelist:
+            if isinstance(x, regions.SkyRegion):
+                regionlist.append(x.to_pixel(self.wcs.celestial))
+            else:
+                regionlist.append(x)
+
+        compound_region = _compound_region(regionlist)
+        mask = compound_region.to_mask()
+
+        xlo, xhi, ylo, yhi = mask.bbox.ixmin, mask.bbox.ixmax, mask.bbox.iymin, mask.bbox.iymax
+
+        # Negative indices will do bad things, like wrap around the cube
+        # If xhi/yhi are negative, there is not overlap
+        if (xhi < 0) or (yhi < 0):
+            raise ValueError("Region is outside of cube.")
+
+        if xlo < 0:
+            xlo = 0
+        if ylo < 0:
+            ylo = 0
+
+        subcube = self.subcube(xlo=xlo, ylo=ylo, xhi=xhi, yhi=yhi)
+
+        if any(dim == 0 for dim in subcube.shape):
+            if allow_empty:
+                warnings.warn("The derived subset is empty: the region does not"
+                              " overlap with the cube (but allow_empty=True).")
+            else:
+                raise ValueError("The derived subset is empty: the region does not"
+                                 " overlap with the cube.")
+
+        maskarray = mask.data[(mask.data.shape[0]-subcube.shape[1]):, (mask.data.shape[1]-subcube.shape[2]):]
```

Comment:
This looks approximately right to me.  Add some comments here, though, to indicate what you're doing so we understand when we read this code later:  you are cropping the mask from the bottom-left because you have just done the same thing with the cube.

We probably also need to crop on the top-right, though, otherwise there's potential for the mask and cube shapes to be mismatched, no?

Reply:
the test that I added is failing. Can you look into that?? I am not able to understand the cause for failure.

Reply:
could you post the traceback?

Reply:
________________________________________________________________ test_ds9region_255_new[255-pixel.reg] ___________________________________________________________________

regfile = '255-pixel.reg'

    @pytest.mark.parametrize('regfile',
                             ('255-fk5.reg', '255-pixel.reg'),
                            )
    def test_ds9region_255_new(regfile):
        # specific test for correctness
        cube, data = cube_and_raw('255.fits')
    
        shapelist = regions.read_ds9(path(regfile))
    
>       subcube = cube.subcube_from_ds9region_new(shapelist)

spectral_cube/tests/test_subcubes.py:104: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
spectral_cube/spectral_cube.py:1975: in subcube_from_ds9region_new
    return masked_subcube.minimal_subcube(spatial_only=True)
spectral_cube/spectral_cube.py:1667: in minimal_subcube
    spatial_only=spatial_only)]
spectral_cube/spectral_cube.py:1706: in subcube_slices_from_mask
    wcs_tolerance=self._wcs_tolerance)
spectral_cube/masks.py:113: in include
    return self._include(data=data, wcs=wcs, view=view)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <spectral_cube.masks.CompositeMask object at 0x7f3538227e10>
data = array([[[11, 12, 13, 14],
        [16, 17, 18, 19],
        [21, 22, 23, 24]],

       [[36, 37, 38, 39],
        [41, 42, 43, 44],
        [46, 47, 48, 49]]])
wcs = WCS Keywords

Number of WCS axes: 3
CTYPE : 'RA---SIN'  'DEC--SIN'  'VOPT'  
CRVAL : 23.1837500515  30.5765277962  -32...0  
PC3_1 PC3_2 PC3_3  : 0.0  0.0  1.0  
CDELT : -0.000555555561268  0.000555555561268  1288.21496879  
NAXIS : 5  5  2
view = ()

    def _include(self, data=None, wcs=None, view=()):
        result_mask_1 = self._mask1._include(data=data, wcs=wcs, view=view)
        result_mask_2 = self._mask2._include(data=data, wcs=wcs, view=view)
        if self._operation == 'and':
>           return result_mask_1 & result_mask_2
E           TypeError: ufunc 'bitwise_and' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

spectral_cube/masks.py:377: TypeError
--------------------------------------------------------------------------- Captured stdout call ----------------------------------------------------------------------------
[[0. 1. 1. 0.]
 [0. 1. 1. 0.]
 [0. 0. 0. 0.]]
(4, 4) (3, 4)

Reply:
it is throwing this type error ,
`TypeError: ufunc 'bitwise_and' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''`

Reply:
It looks like the `maskarray` needs to be converted to a boolean array...

Reply:
so, on this line, do `maskarray = (...).astype('bool')`

Reply:
Thanks. also,I will add some comments and wrt the cropping I will think a little bit more and come back to you .